### PR TITLE
ARROW-7948: [Go] Decimal128 Integration fix

### PR DESCRIPTION
**Pull request by zeroshade**
Had to redesign a bit of the arrjson handling to do this properly, but I think it's better this way anyways.

### CODE REVIEW

**Comment by zeroshade** [comment]
tagging @pitrou since the issue ARROW-7948 was originally filed by him.

CC @sbinet @emkornfield

File: dev/archery/archery/integration/datagen.py [hunk 1]
```diff
@@ -1527,7 +1527,6 @@ def _temp_path():
         .skip_category('Go'),  # TODO(ARROW-7901)
 
         generate_decimal128_case()
-        .skip_category('Go')  # TODO(ARROW-7948): Decimal + Go
```
Comment (emkornfield) [inline dev/archery/archery/integration/datagen.py:1530]: +1

File: go/arrow/decimal128/decimal128.go [hunk 2]
```diff
@@ -54,6 +59,36 @@ func FromI64(v int64) Num {
...
+	v.FillBytes(buf[:])
+	return Num{
+		lo: binary.BigEndian.Uint64(buf[8:]),
+		hi: int64(binary.BigEndian.Uint64(buf[:8])),
+	}
+}
+
+func FromBigInt(v *big.Int) Num {
+	if v.Sign() < 0 {
+		// if the value is negative, then get the high and low bytes from the
```
Comment (emkornfield) [inline go/arrow/decimal128/decimal128.go:76]: sorry not sure I follow why the high/slow byte approach doesn't work for negative values?

File: go/arrow/decimal128/decimal128.go [hunk 3]
```diff
@@ -54,6 +59,36 @@ func FromI64(v int64) Num {
 	}
 }
 
+func fromBigIntPositive(v *big.Int) Num {
```
Comment (emkornfield) [inline go/arrow/decimal128/decimal128.go:62]: Big int seems like it is arbitrary precision?  What happens if the value exceeds 16 bytes?

**Comment by emkornfield** [comment]
Two questions, otherwise looks reasonable to me.

File: go/arrow/decimal128/decimal128.go (hunk 2 above)
Comment (zeroshade) [reply go/arrow/decimal128/decimal128.go]: updated with more clear comments, but essentially: big.Int stores the value as a boolean for the sign + absolute value, arrow decimal128 expects high and low bytes as 2's compliment values. So in order to have the correct value we need to take the bytes (absolute value) and then negate via 2's compliment if it was negative.

File: go/arrow/decimal128/decimal128.go (hunk 3 above)
Comment (zeroshade) [reply go/arrow/decimal128/decimal128.go]: added a check and panic if `v.BitLen()` > 128

**Comment by zeroshade** [comment]
the slight changes in the updated commits significantly improve the performance which can be seen by the benchmarks.

Before:
```
goos: windows
goarch: amd64
pkg: github.com/apache/arrow/go/arrow/decimal128
cpu: Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz
BenchmarkBigIntToDecimal-12    	20276983	        58.21 ns/op	      48 B/op	       1 allocs/op
BenchmarkDecimalToBigInt-12    	 8945876	       141.2 ns/op	     112 B/op	       4 allocs/op
PASS
ok  	github.com/apache/arrow/go/arrow/decimal128	2.839s
```
After:
```
goos: windows
goarch: amd64
pkg: github.com/apache/arrow/go/arrow/decimal128
cpu: Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz
BenchmarkBigIntToDecimal-12    	262177144	         4.549 ns/op	       0 B/op	       0 allocs/op
BenchmarkDecimalToBigInt-12    	21051966	        58.65 ns/op	      48 B/op	       2 allocs/op
PASS
ok  	github.com/apache/arrow/go/arrow/decimal128	3.045s
```

**Comment by zeroshade** [comment]
The Dev/Lint job failure doesn't appear to be caused by anything in this change as far as i can tell.

**Comment by emkornfield** [comment]
+1.  Thank you.  @zeroshade not sure if you are interested but I don't think Go supports Decimal256 yet.

**Comment by zeroshade** [comment]
@emkornfield It doesn't, that's on my list of stuff that i plan on making PRs for haha. 😄

**Pull request closed without merging**